### ISSUE TITLE
Allow scripts to export static variables, and allow them to be assigned in the editor

### ISSUE THREAD

**Issue by pineapplemachine** [godotengine/godot-proposals#10961]
### Describe the project you are working on

Currently I am working on a project where it would be very helpful for certain node or resource types to have a globally configured reference to another value or resource, without having to hardcode paths, or having to manually specify the reference for every instance.

### Describe the problem or limitation you are having in your project

There is seemingly no easy way to create and handle custom global configuration values in Godot.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

This would become allowed, e.g. in GDScript:

```gdscript
class_name MyConfigurableType

@export static var my_resource: MyResourceType
```

When an object of this type is viewed in the inspector, an additional section titled e.g. "MyConfigurableType Static" would show alongside the instance sections. Here, the global values would be visible and configurable.

Ideally, there should also be a new tab or window which aggregates all of these static exports in one place, without needing to create an instance of MyConfigurableType in order to view and edit the global values.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

In GDScript:

```gdscript
@export static var my_resource: MyResourceType
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

There does not seem to be any existing clean solution for this.

### Is there a reason why this should be core and not an add-on in the asset library?

This would involve a GDScript language change.

**Comment by AThousandShips** [comment]
Having non-instance data be editable in the inspector is IMO confusing and error prone, it also doesn't really make sense UI wise to me, having it be configurable in a dedicated project settings menu would make more sense

The bigger question is how it would work, your proposal doesn't actually explain how this would be loaded and initialized, or where it would be stored, or how it would handle values between the class and such configuration 

What happens if you assign a value in the script source? Is it the default? What happens if you change this? Is the configured value in the settings/inspector used?

**Comment by dalexeev** [comment]
Static variables belong to the class, not to instances. I don't think this is a good idea, since it's not clear what to do with multiple instances exporting different values ​​of static variables and when they should be applied.

You could store global parameters in a separate resource instead:

```gdscript

const PARAMS = preload("./params.tres")

static var param_1 = PARAMS.param_1
static var param_2 = PARAMS.param_2
```

It's even [possible](https://github.com/dalexeev/gdscript-compile-time-evaluations) to use them as constants, although I'm not sure that's the intention:

```gdscript

const PARAMS = preload("./params.tres")

const PARAM_1 = PARAMS.param_1
const PARAM_2 = PARAMS.param_2
```

**Comment by pineapplemachine** [comment]
@AThousandShips I am not knowledgeable enough about engine internals to suggest an implementation.

But I think assigning the value in code would work like assigning any exported instance var in code. At runtime, it would update the value like any typical mutable global value. In a tool script, it would make the change persistently, and would be reflected in whatever editor UI is used to assign these values.

@dalexeev In this case, the example `static var` and `const` declarations cannot be assigned in the editor, only in code, and things like scene paths will not be automatically updated as they are with exported instance vars. I don't think this solves the usability problem.

One possible way to handle this is to create a custom Resource class, create a single instance somewhere in the project directory, and add a `static func` to the Resource class for getting that single designated instance. This is what I've ended up doing in some cases where I needed globally configured references to resources (e.g. which Texture2D to use in the UI in places where an icon is meant to represent some game concept) that ideally I wanted to 1. edit in the inspector rather than in code and 2. have paths automatically update and be repaired in case I moved or renamed a resource, rather than silently broken.

But global custom project settings, which seems like something with almost universal utility, should really should have some option with better usability in my opinion.

**Comment by MrGreaterThan** [comment]
> it's not clear what to do with multiple instances exporting different values ​​of static variables and when they should be applied.

I don't agree. You won't have "multiple instances exporting different values" because all instances would share a single static value. That's all a static member is.

I imagine it would work like this:
- Exported static variables have some visual distinction in the inspector. (Different color or grouped into a section together)
- Changing the value in the inspector of one instance also changes it across every other instance.
- The inspector value gets assigned to the static member when the game is run.

I imagine the values for export variables are stored in the scene files somewhere. If that's the case, static export variable settings would have to be stored differently somewhere at the project-level. I think assigning all of your would-be static references in a singleton/autoload is a good enough workaround so if this is difficult to implement, it's a feature I can live without.

**Comment by Pheubel** [comment]
To me it would make sense that if an export is shared across class instanced, that it would be accessible through ProjectSettings -> Globals. but i am not sure if that's feasable

**Comment by Hinara** [comment]
Another way might be to expose those fields on the script itself, not on the instances, any time the script is loaded the values defined in the editor would be set to the one set in the editor. 

This also might allow global scripts to have a way to have some fields set via the inspector allowing to have explicit dependencies.

However this might require some metadata files to hold those values alongside the script.

**Comment by ComplianceFailure** [comment]
I would like for this to be implemented in some capacity. As the user above said, the inspector field for the script file would make perfect sense as an alternative to static vars being listed among instance data in the inspector.

There are definitely ways around this and they work perfectly well for most use cases.
However, I need my game to be able to step its "main loop" which involves simulating the motion and faux-physics of a large quantity of objects multiple times between frames. I believe minimizing property lookups is important in enabling this, but that currently means sacrificing the ability to tweak parameters in the inspector at runtime.

For anyone reading this in the future and this is still unresolved, you can override `Node._get_property_list` and friends to forward static variables of your choice to the inspector.

**Comment by Naros** [comment]
@ComplianceFailure doesn't this break down when you attach the same script to two different nodes in the scene, each where you specify this static variable's value differently? I certainly would expect that likely doesn't work as you might think it does in all use cases.

**Comment by ComplianceFailure** [comment]
> [@ComplianceFailure](https://github.com/ComplianceFailure) doesn't this break down when you attach the same script to two different nodes in the scene, each where you specify this static variable's value differently? I certainly would expect that likely doesn't work as you might think it does in all use cases.

Yes, if treated as typical property overrides of the object saved to the scene, the behavior would be unintuitive.

This theoretical feature should treat these fields differently by disallowing per-instance overriding and instead directly read/write to a script's metadata somewhere. Admittedly, I've got no idea where that information would be saved, but I have to imagine there is some place that makes sense.

Solving that, this would effectively allow you to view the values of, override the initializer of, and set at runtime the values of any static variable belonging to a script or custom class by accessing an inspector field for the class itself, nested within the associated GDScript Resource's inspector field.

Of course, there are surely better ways to go about this. Since the variables can be accessed in code via `object_name.static_var_name`, it would make also sense for a static variable declared with `@export` to be listed among the object's other properties.

**Comment by bluebirdsan** [comment]
I think this would be a good feature to have.

I'm making a sort of string diff checker and would like to have colors to highlight changes and it'd be nice to be able to set the color with the inspectors color selector.

Having static variables show up and be set on the script's inspector (as Hinara and ComplianceFailure said) would be a good place, because the variable would show up in only one place and not on every instance. But it'd also be convenient to have static vars on instances, but changing them would change the static var on every instance.
